File: android/src/main/java/com/audiobrowser/util/PlayingStateFactory.kt
package com.audiobrowser.util

import com.margelo.nitro.audiobrowser.PlaybackState
import com.margelo.nitro.audiobrowser.PlayingState

object PlayingStateFactory {
  /**
   * Derives a PlayingState from playWhenReady and state.
   *
   * @param playWhenReady Whether the player wants to play when ready
   * @param playbackState The current player state
   * @return A PlayingState representing the current playing/buffering status
   */
  fun derive(playWhenReady: Boolean, playbackState: PlaybackState): PlayingState {
    val playing =
      playWhenReady &&
        !(playbackState == PlaybackState.ERROR ||
          playbackState == PlaybackState.ENDED ||
          playbackState == PlaybackState.NONE)
    val buffering =
      playWhenReady &&
        (playbackState == PlaybackState.LOADING || playbackState == PlaybackState.BUFFERING)
    return PlayingState(playing, buffering)
  }
}

File: src/web/TrackPlayer/PlayingStateFactory.ts
import type { PlaybackState, PlayingState } from '../../features'

/**
 * Derives the two-flag {@link PlayingState} (playing / buffering) from the
 * player's intent (`playWhenReady`) and its current {@link PlaybackState}.
 *
 * A faithful port of the native derivations (android `PlayingStateFactory.derive`,
 * ios `PlayingStateManager.update`) so all platforms agree on what "playing" and
 * "buffering" mean for the now-playing UI.
 */
export function derivePlayingState(
  playWhenReady: boolean,
  state: PlaybackState
): PlayingState {
  return {
    playing:
      playWhenReady &&
      state !== 'error' &&
      state !== 'ended' &&
      state !== 'none',
    buffering: playWhenReady && (state === 'loading' || state === 'buffering')
  }
}
